### PR TITLE
resolve: diagnostics improvement and groundwork for RFC 1560

### LINKED ISSUE #35115: resolve: "bindings cannot shadow constants" diagnostics are unreliable
```rust
const a: u8 = 0; //~ NOTE constant `a` is defined here

fn main() {
    // fn f() {} // If this line is uncommented, the `NOTE` is suppressed.
    let a = 0; //~ ERROR let bindings cannot shadow constants
}

### LINKED ISSUE #35135: resolve: suboptimal unused import detection
For example:
```rust
pub mod foo {
    pub fn bar() {}
}

use foo::bar; // This should be an unused import warning

fn main() {
    let bar = 0;
    println!("{}", bar);
}
```

### PR DESCRIPTION
**Pull request by jseyfried**
Fixes #35115, fixes #35135, and lays groundwork for #32213 (cc #35120).
r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @petrochenkov

File: src/librustc_resolve/resolve_imports.rs [hunk 1]
```diff
@@ -343,6 +343,25 @@ struct ImportResolver<'a, 'b: 'a> {
     resolver: &'a mut Resolver<'b>,
 }
 
+impl<'a, 'b: 'a> ::std::ops::Deref for ImportResolver<'a, 'b> {
+    type Target = Resolver<'b>;
+    fn deref(&self) -> &Resolver<'b> {
+        self.resolver
+    }
+}
```
Comment (nrc) [inline src/librustc_resolve/resolve_imports.rs]: I feel dirty just reviewing this code - I really wish we didn't use Deref like this in name resolution. Still the damage is done, I guess this doesn't make things much worse.

**Comment by sanxiyn** [comment]
cargotest failure is legit. (This PR produced additional unused import warnings in cargo which is denied.)

**Comment by sanxiyn** [comment]
FYI, you need to explicitly update commit hash in `src/tools/cargotest/main.rs`.

**Comment by jseyfried** [comment]
@sanxiyn Thanks -- didn't know that!
I updated the `cargo_test` hash in the above commit.

@bors r=nrc

**Comment by jseyfried** [comment]
(pending https://github.com/rust-lang/cargo/pull/2960)

**Pull request merged**